homeassistant/helpers/trigger.py

```diff
@@ -101,20 +107,51 @@ async def async_validate_trigger_config(
 def _trigger_action_wrapper(
     hass: HomeAssistant, action: Callable, conf: ConfigType
 ) -> Callable:
-    """Wrap trigger action with extra vars if configured."""
+    """Wrap trigger action with extra vars if configured.
+
+    If action is a coroutine function, a coroutine function will be returned.
+    If action is a callback, a callback will be returned.
+    """
     if CONF_VARIABLES not in conf:
         return action
 
-    @functools.wraps(action)
-    async def with_vars(
-        run_variables: dict[str, Any], context: Context | None = None
-    ) -> None:
-        """Wrap action with extra vars."""
-        trigger_variables = conf[CONF_VARIABLES]
-        run_variables.update(trigger_variables.async_render(hass, run_variables))
-        await action(run_variables, context)
+    # Check for partials to properly determine if coroutine function
+    check_func = action
+    while isinstance(check_func, functools.partial):
+        check_func = check_func.func
```

Comment:
Not needed if you use [`inspect.iscoroutinefunction`](https://docs.python.org/3/library/inspect.html#inspect.iscoroutinefunction)

Reply:
While it's not needed for `inspect.iscoroutinefunction`, it's needed for `is_callback`.
If we teach `is_callback` about unwinding partials we can indeed drop it, that's for another PR though.